rudder-core/src/main/scala/com/normation/rudder/services/policies/write/Cf3PromisesFileWriterService.scala

```diff
@@ -0,0 +1,602 @@
+
+        for (folder <- newFolders) {
+          logger.info("Restoring old promises on folder %s".format(folder.baseFolder))
+          try {
+            restoreBackupNodeFolder(folder.baseFolder, folder.backupFolder);
+          } catch {
+            case ex: Exception =>
+              logger.error("could not restore old promises into %s ".format(folder.baseFolder))
+              throw ex
+          }
+        }
+        throw ex
+    }
+
+  }
+
+
+
+  /**
+   * Move the machine promises folder  to the backup folder
+   * @param machineFolder
+   * @param backupFolder
+   */
+  private[this] def backupNodeFolder(nodeFolder: String, backupFolder: String): Unit = {
+    val src = new File(nodeFolder)
+    if (src.isDirectory()) {
+      val dest = new File(backupFolder)
+      if (dest.isDirectory)
+        // force deletion of previous backup
+        FileUtils.forceDelete(dest)
+
+      FileUtils.moveDirectory(src, dest)
+    }
+  }
+
+  /**
+   * Move the newly created folder to the final location
+   * @param newFolder : where the promises have been written
+   * @param nodeFolder : where the promises will be
+   */
+  private[this] def moveNewNodeFolder(sourceFolder: String, destinationFolder: String): Unit = {
+    val src = new File(sourceFolder)
+
+    logger.debug("Moving folders from %s to %s".format(src, destinationFolder))
+
+    if (src.isDirectory()) {
+      val dest = new File(destinationFolder)
+
+      if (dest.isDirectory)
+        // force deletion of previous promises
+        FileUtils.forceDelete(dest)
+
+      FileUtils.moveDirectory(src, dest)
+
+      // force deletion of dandling new promise folder
+      if ( (src.getParentFile().isDirectory) && (src.getParent().endsWith("rules.new")))
+        FileUtils.forceDelete(src.getParentFile())
+
+    } else {
+      logger.error("Could not find freshly created promises at %s".format(sourceFolder))
+      throw new IOException("Created promises not found !!!!")
+    }
+  }
+
+  /**
+   * Restore (by moving) backup folder to its original location
+   * @param machineFolder
+   * @param backupFolder
+   */
+  private[this] def restoreBackupNodeFolder(nodeFolder: String, backupFolder: String): Unit = {
+    val src = new File(backupFolder)
+    if (src.isDirectory()) {
+      val dest = new File(nodeFolder)
+      // force deletion of invalid promises
+      FileUtils.forceDelete(dest)
+
+      FileUtils.moveDirectory(src, dest)
+    } else {
+      logger.error("Could not find freshly backup promises at %s".format(backupFolder))
+      throw new IOException("Backup promises could not be found, and valid promises couldn't be restored !!!!")
+    }
+  }
+
+
+  /**
+   * Calculate path for node configuration.
+   * Path are agent dependant, so from that point, all
+   * node configuration are also identified by agent.
+   * Note that a node without any agent configured won't
+   * have any promise written.
+   */
+  private[this] def calculatePathsForNodeConfigurations(
```

Comment:
the order of the method in the file is a bit off... i guess they'd be found more easily if htey were either in chronological order, or utiliies methods at the end, but here it looks a bit random